modules/core/src/main/scala/com.snowplowanalytics.snowplow.bigquery/processing/LegacyColumns.scala

```diff
@@ -0,0 +1,247 @@
+    entityType: TabledEntity.EntityType,
+    failure: FailureDetails.LoaderIgluError
+  )
+
+  def resolveTypes[F[_]: Sync: RegistryLookup](
+    resolver: Resolver[F],
+    entities: Map[TabledEntity, Set[SchemaSubVersion]]
+  ): F[Result] =
+    entities.toVector
+      .flatMap { case (tabledEntity, subVersions) =>
+        subVersions.map { subVersion =>
+          (tabledEntity.entityType, TabledEntity.toSchemaKey(tabledEntity, subVersion))
+        }
+      }
+      .traverse { case (entityType, schemaKey) =>
+        getSchema(resolver, schemaKey)
+          .map { schema =>
+            val mode = entityType match {
+              case TabledEntity.UnstructEvent => LegacyMode.Nullable
+              case TabledEntity.Context       => LegacyMode.Repeated
+            }
+            val columnName  = legacyColumnName(entityType, schemaKey)
+            val legacyField = LegacyField.build(columnName, schema, false).setMode(mode).normalized
+            val v2Field     = legacyFieldToV2Field(legacyField)
+            FieldForEntity(v2Field, schemaKey, entityType)
+          }
+          .leftMap(ColumnFailure(schemaKey, entityType, _))
+          .value
+      }
+      .map { eithers =>
+        val (failures, good) = eithers.separate
+        Result(good, failures.toList)
+      }
+
+  def transformEvent(
+    processor: BadRowProcessor,
+    event: Event,
+    batchInfo: Result,
+    loadTstamp: java.lang.Long
+  ): Either[BadRow, Map[String, AnyRef]] =
+    for {
+      _ <- failForResolverErrors(processor, event, batchInfo.igluFailures)
+      atomic <- forAtomic(processor, event)
+      nonAtomic <- forEntities(event, batchInfo.fields).leftMap { nel =>
+                     BadRow.LoaderIgluError(processor, BadRowFailure.LoaderIgluErrors(nel), BadPayload.LoaderPayload(event))
+                   }.toEither
+    } yield atomic ++ nonAtomic + ("load_tstamp" -> loadTstamp)
+
+  private def legacyFieldToV2Field(legacyField: LegacyField): V2Field = {
+    val fieldType = legacyTypeToV2Type(legacyField.fieldType)
+    legacyField.mode match {
+      case LegacyMode.Nullable =>
+        V2Field(legacyField.name, fieldType, V2Type.Nullability.Nullable)
+      case LegacyMode.Required =>
+        V2Field(legacyField.name, fieldType, V2Type.Nullability.Required)
+      case LegacyMode.Repeated =>
+        val repeatedType = V2Type.Array(fieldType, V2Type.Nullability.Nullable)
+        V2Field(legacyField.name, repeatedType, V2Type.Nullability.Nullable)
+    }
+  }
+
+  private def legacyTypeToV2Type(legacyType: LegacyType): V2Type =
+    legacyType match {
+      case LegacyType.String    => V2Type.String
+      case LegacyType.Boolean   => V2Type.Boolean
+      case LegacyType.Integer   => V2Type.Long
+      case LegacyType.Float     => V2Type.Double
+      case LegacyType.Numeric   => V2Type.Decimal(V2Type.DecimalPrecision.Digits38, 9)
+      case LegacyType.Date      => V2Type.Date
+      case LegacyType.DateTime  => V2Type.Timestamp
```

Comment:
Does it mean that now for v1 there might be somewhere column with `DateType` type? If it's possible, is such `DateTime` -> `Timestamp` conversion safe?

Reply:
From looking [at schema-ddl for BigQuery](https://github.com/snowplow/schema-ddl/blob/release/0.23.0/modules/core/src/main/scala/com.snowplowanalytics/iglu.schemaddl/bigquery/Suggestion.scala) I think the `DateTime` type was never used.

If there is a `DateTime` column out there in the wild.....  For `Timestamp` data we [provide a Long representing microseconds from the epoch](https://github.com/snowplow-incubator/snowplow-bigquery-loader/blob/2.0.0-rc11/modules/core/src/main/scala/com.snowplowanalytics.snowplow.bigquery/processing/BigQueryCaster.scala#L34).  I think this is also compatible with DateTime column, but yeah we should check that.